File: Java Presidents Project/app/src/main/java/com/example/javapresidents/MainActivity.java
//****************************************************************************
package com.example.javapresidents;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity
{
    Button btnAddOne;
    Menu menu;

    MyJavaClass myApplication = (MyJavaClass) this.getApplication();
    List<President> presidentListMain;


    private RecyclerView recyclerView;
    private RecyclerView.Adapter mAdapter;
    private RecyclerView.LayoutManager layoutManager;

    //************************************************************************

    /**@methodName:
     * @param:
     * @return:
     * @description:
     * */

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        presidentListMain = myApplication.getPresidentList();

        btnAddOne = findViewById(R.id.btn_addOne);
        btnAddOne.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View view)
            {
                Intent intent = new Intent(MainActivity.this, AddEditOne.class);
                startActivity(intent);
            } // End onClick
        }); // End btnAddOne.setOnClickListener

        recyclerView = findViewById(R.id.rv_PresdidentList);
        recyclerView.setHasFixedSize(true);

        layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        mAdapter = new RecycleViewAdapter(presidentListMain, MainActivity.this);
        recyclerView.setAdapter(mAdapter);
    } // End onCreate

    //************************************************************************

    /**@methodName:
     * @param:
     * @return:
     * @description:
     * */

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.sort_menu, menu);

        return true;
    } // End onCreateOptionsMenu

    //************************************************************************

    /**@methodName:
     * @param:
     * @return:
     * @description:
     * */

    // Sort presidents
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item)
    {
        switch (item.getItemId())
        {
            // Sort by ascending name
            case R.id.menu_AtoZ:
                Collections.sort(presidentListMain,
                                 President.PresidentNameAZComparator);
                Toast.makeText(getApplicationContext(), "Sorted Name Ascending",
                               Toast.LENGTH_SHORT).show();
                mAdapter.notifyDataSetChanged();
                return true;

            // Sort by descending name
            case R.id.menu_ZtoA:
                Collections.sort(presidentListMain,
                                 President.PresidentNameZAComparator);
                Toast.makeText(getApplicationContext(),
                               "Sorted Name Descending", Toast.LENGTH_SHORT)
                     .show();
                mAdapter.notifyDataSetChanged();
                return true;

            // Sort by ascending date
            case R.id.menu_dateAscending:
                Collections.sort(presidentListMain,
                                 President.PresidentDateAscendingComparator);
                Toast.makeText(getApplicationContext(), "Sorted Date Ascending",
                               Toast.LENGTH_SHORT).show();
                mAdapter.notifyDataSetChanged();
                return true;

            // Sort by descending date
            case R.id.menu_dateDescending:
                Collections.sort(presidentListMain,
                                 President.PresidentDateDescendingComparator);
                Toast.makeText(getApplicationContext(),
                               "Sorted Date Descending", Toast.LENGTH_SHORT)
                     .show();
                mAdapter.notifyDataSetChanged();
                return true;
        } // End switch
        return super.onOptionsItemSelected(item);
    } // End onOptionsItemSelected
} // End MainActivity class

File: app/src/main/java/com/example/presidentlist/MainActivity.kt
package com.example.presidentlist

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.*
import java.util.Collections.*


class MainActivity : AppCompatActivity()
{
    var menu: Menu? = null
    private var presidentList: MutableList<President>? = null
    lateinit var recyclerView: RecyclerView
    private var mAdapter: RecyclerView.Adapter<*>? = null
    
    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    
        // Add one button and intent
        val btnAddOne: Button = findViewById(R.id.btn_addOne)
        val addEditOneIntent = Intent()
        addEditOneIntent.setClass(this, AddEditOne::class.java)
        
        // Initialize list and adapter
        presidentList = MyKotlinClass.getPresidentList()
        mAdapter = RecycleViewAdapter(presidentList!!, this)
    
        // Recycler View population and padding
        recyclerView = findViewById(R.id.rv_PresdidentList)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = mAdapter
        
        // Check president count
        Toast.makeText(this, "President Count: ${presidentList!!.size}",
            Toast.LENGTH_SHORT).show()
        
        // Go to Add Edit One Activity on click
        btnAddOne.setOnClickListener {
            startActivity(addEditOneIntent)
        } // End btnAddOne.setOnClickListener
    } // End onCreate
    
    //************************************************************************
    
    /**@methodName:
     * @param:
     * @return:
     * @description:
     */
    
    override fun onCreateOptionsMenu(menu: Menu): Boolean
    {
        val inflater = menuInflater
        inflater.inflate(R.menu.sort_menu, menu)
        return true
    } // End onCreateOptionsMenu
    
    //************************************************************************
    /**@methodName: onOptionsItemSelected
     * @param: item (MenuItem)
     * @return: Bool (t/f if the presidents could be sorted.)
     * @description: Sorts presidents
     */
    
    @SuppressLint("NotifyDataSetChanged")
    override fun onOptionsItemSelected(item: MenuItem): Boolean
    {
        when (item.itemId)
        {
            // Ascending Order
            R.id.menu_AtoZ ->
            {
                presidentList?.let {
                    sort(it, President.PresidentNameAZComparator)
                }
                Toast.makeText(applicationContext, "Ascending",
                    Toast.LENGTH_SHORT).show()
                mAdapter!!.notifyDataSetChanged()
                return true
            } // End R.id.menu_AtoZ
            
            // Descending
            R.id.menu_ZtoA ->
            {
                presidentList?.let {
                    sort(it, President.PresidentNameZAComparator)
                }
                Toast.makeText(applicationContext, "Descending",
                    Toast.LENGTH_SHORT).show()
                mAdapter!!.notifyDataSetChanged()
                return true
            } // End R.id.menu_ZtoA
            
            // Oldest
            R.id.menu_dateAscending ->
            {
                presidentList?.let {
                    sort(it, President.PresidentDateAscendingComparator)
                }
                Toast.makeText(applicationContext, "Chronological",
                    Toast.LENGTH_SHORT).show()
                mAdapter!!.notifyDataSetChanged()
                return true
            } // End R.id.menu_dateAscending
            
            // Most Recent
            R.id.menu_dateDescending ->
            {
                presidentList?.let {
                    sort(it, President.PresidentDateDescendingComparator)
                }
                Toast.makeText(applicationContext, "Most Recent",
                    Toast.LENGTH_SHORT).show()
                mAdapter!!.notifyDataSetChanged()
                return true
            } // End R.id.menu_dateDescending
        } // End when
        return super.onOptionsItemSelected(item)
    } // End onOptionsItemSelected
} // End MainActivity class
